File: app/src/main/java/vn/vistark/giam_sat_nha_yen/ui/dashboard_screen/DashboardScreenActivity.java
package vn.vistark.giam_sat_nha_yen.ui.dashboard_screen;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.os.SystemClock;
import android.util.Log;
import android.view.SurfaceView;
import android.view.View;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SwitchCompat;
import androidx.recyclerview.widget.RecyclerView;

import org.opencv.android.BaseLoaderCallback;
import org.opencv.android.CameraBridgeViewBase;
import org.opencv.android.LoaderCallbackInterface;
import org.opencv.android.OpenCVLoader;
import org.opencv.android.Utils;
import org.opencv.core.CvType;
import org.opencv.core.Mat;

import es.dmoral.toasty.Toasty;
import vn.vistark.giam_sat_nha_yen.R;
import vn.vistark.giam_sat_nha_yen.data.arduino_community.ArduinoCommunity;
import vn.vistark.giam_sat_nha_yen.data.firebase.FirebaseConfig;
import vn.vistark.giam_sat_nha_yen.data.modal.Auto;
import vn.vistark.giam_sat_nha_yen.ui.BaseAppCompatActivity;
import vn.vistark.giam_sat_nha_yen.ui.dashboard_screen.timer_dialog.TimerDialog;
import vn.vistark.giam_sat_nha_yen.ui.dashboard_screen.video_transfer.VideoTransfer;
import vn.vistark.giam_sat_nha_yen.utils.ScreenUtils;

public class DashboardScreenActivity extends AppCompatActivity implements BaseAppCompatActivity, CameraBridgeViewBase.CvCameraViewListener2 {
    private final static String TAG = DashboardScreenActivity.class.getSimpleName();

    // Danh sách các View sẽ được ánh xạ
    private ImageView ivAppPowerOffButton;
    private RecyclerView rvTimerList;
    private Button btnAddNewTimer;
    private TextView tvCurrentTime;
    private TextView tvEstablishTime;
    private TextView tvHumidity;
    private TextView tvTemperature;
    private TextView tvTimerListEmptyNotify;
    private TextView tvAutoTemp;
    private TextView tvAutoHumidity;
    private SwitchCompat scTimerAutoPower;

    private LinearLayout lnMultiCameraButton;
    private LinearLayout lnMultiCamLayout;
    private ImageView ivImageCapturedS1;
    private ImageView ivImageCapturedS2;
    private ImageView ivImageCapturedS3;
    private ImageView ivImageCapturedS4;

    // Loads camera view of OpenCV for us to use. This lets us see using OpenCV
    private CameraBridgeViewBase mOpenCvCameraView;
    private ImageView ivImageCaptured;

    // These variables are used (at the moment) to fix camera orientation from 270degree to 0degree
    Mat mRgba;
    Mat mRgbaF;
    Mat mRgbaT;


    private DashboardScreenPresenter mPresenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_dashboard_screen);
        ScreenUtils.hideTitleBarAndTransparentStatusBar(this);
        initViews();
        initPresenters();
        initEvents();
        initAutoSync();
        loadTimerList();
        initCamera();
        ArduinoCommunity.asyncCommunity(this);
    }

    private void initAutoSync() {
        new Auto().sync(FirebaseConfig.myRef, this);
    }

    private BaseLoaderCallback mLoaderCallback = new BaseLoaderCallback(this) {
        @Override
        public void onManagerConnected(int status) {
            if (status == LoaderCallbackInterface.SUCCESS) {
                mOpenCvCameraView.enableView();
                showMultiCamera();
            } else {
                super.onManagerConnected(status);
            }
        }
    };

    private void initCamera() {
        this.mOpenCvCameraView.setVisibility(SurfaceView.VISIBLE);
        this.mOpenCvCameraView.setCvCameraViewListener(this);
    }

    private void loadTimerList() {
        mPresenter.loadTimerList(rvTimerList);
    }

    @Override
    public void initViews() {
        ivAppPowerOffButton = findViewById(R.id.appPowerOff);
        rvTimerList = findViewById(R.id.timerList);
        btnAddNewTimer = findViewById(R.id.addNewTimerBtn);
        mOpenCvCameraView = (CameraBridgeViewBase) findViewById(R.id.imageCaptureFrame);
        ivImageCaptured = findViewById(R.id.imageCaptured);
        tvCurrentTime = findViewById(R.id.currentTime);
        tvEstablishTime = findViewById(R.id.establishTime);
        tvHumidity = findViewById(R.id.humidity);
        tvTemperature = findViewById(R.id.temperature);
        tvTimerListEmptyNotify = findViewById(R.id.timerListEmptyNotify);
        lnMultiCameraButton = findViewById(R.id.multiCameraButton);
        lnMultiCamLayout = findViewById(R.id.multiCamLayout);
        ivImageCapturedS1 = findViewById(R.id.imageCapturedS1);
        ivImageCapturedS2 = findViewById(R.id.imageCapturedS2);
        ivImageCapturedS3 = findViewById(R.id.imageCapturedS3);
        ivImageCapturedS4 = findViewById(R.id.imageCapturedS4);
        tvAutoTemp = findViewById(R.id.autoTemp);
        tvAutoHumidity = findViewById(R.id.autoHumidity);
        scTimerAutoPower = findViewById(R.id.timerAutoPower);
    }

    public void showMultiCamera() {
        lnMultiCameraButton.post(new Runnable() {
            @Override
            public void run() {
                lnMultiCameraButton.setVisibility(View.GONE);
                ivImageCaptured.setVisibility(View.GONE);
                lnMultiCamLayout.setVisibility(View.VISIBLE);
            }
        });
    }

    public void chooseShowCam1() {
        ivImageCapturedS1.post(new Runnable() {
            @Override
            public void run() {
                lnMultiCameraButton.setVisibility(View.VISIBLE);
                ivImageCaptured.setVisibility(View.VISIBLE);
                lnMultiCamLayout.setVisibility(View.GONE);
            }
        });
    }

    @Override
    public void initEvents() {
        // Bắt sự kiện khi nhấn vào nút đóng
        ivAppPowerOffButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ivAppPowerOffButton.post(new Runnable() {
                    @Override
                    public void run() {
                        mPresenter.powerOffConfirm();
                    }
                });
            }
        });

        // Bắt sự kiện khi nhấn vào nút thêm mới hẹn giờ
        btnAddNewTimer.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                TimerDialog.showAdd(DashboardScreenActivity.this);
            }
        });

        // Bắt sự kiện khi nhấn vào nút nhiều camera
        lnMultiCameraButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                showMultiCamera();
            }
        });

        // Sự kiện khi nhấn vào 1 khung hình của cam bất kỳ trong chế độ xem nhiều camera
        // Nhấn màn hình 1
        ivImageCapturedS1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                chooseShowCam1();
            }
        });
        // Nhấn màn hình 2,3,4
        ImageView[] ivs = {ivImageCapturedS2, ivImageCapturedS3, ivImageCapturedS4};
        for (final ImageView v : ivs) {
            v.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    v.post(new Runnable() {
                        @Override
                        public void run() {
                            Toasty.info(DashboardScreenActivity.this, "Không nhận được tín hiệu từ camera này!", Toasty.LENGTH_SHORT, false).show();
                        }
                    });
                }
            });
        }

        // Sự kiện cho nút nguồn auto
        scTimerAutoPower.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                mPresenter.changeListByStateOfAutoPower(b);
            }
        });
    }


    @Override
    public void initPresenters() {
        mPresenter = new DashboardScreenPresenter(this);
    }

    @Override
    public void onBackPressed() {
        mPresenter.powerOffConfirm();
    }

    public void updateAutoHumidity(final String start, final String end) {
        tvAutoHumidity.post(new Runnable() {
            @SuppressLint("SetTextI18n")
            @Override
            public void run() {
                tvAutoHumidity.setText(start + " - " + end + "%");
            }
        });
    }

    public void updateAutoTemp(final String start, final String end) {
        tvAutoTemp.post(new Runnable() {
            @SuppressLint("SetTextI18n")
            @Override
            public void run() {
                tvAutoTemp.setText(start + " - " + end);
            }
        });
    }

    public boolean getStateAutoPower() {
        return scTimerAutoPower.isChecked();
    }

    public void updateTimerAutoPower(final Boolean state) {
        scTimerAutoPower.post(new Runnable() {
            @Override
            public void run() {
                scTimerAutoPower.setChecked(state);
            }
        });
    }

    public void updateHumidityView(final String humidity) {
        tvHumidity.post(new Runnable() {
            @SuppressLint("SetTextI18n")
            @Override
            public void run() {
                tvHumidity.setText(humidity + "%");
            }
        });
    }

    public void updateTemperature(final String temperature) {
        tvTemperature.post(new Runnable() {
            @SuppressLint("SetTextI18n")
            @Override
            public void run() {
                tvTemperature.setText(temperature + "°C");
            }
        });
    }

    public void updateCurrentTimeView(final String currentTime) {
        tvCurrentTime.post(new Runnable() {
            @Override
            public void run() {
                tvCurrentTime.setText(currentTime);
            }
        });
    }

    public void updateEstablishTimeView(final String establedTime) {
        tvEstablishTime.post(new Runnable() {
            @Override
            public void run() {
                tvEstablishTime.setText(establedTime);
            }
        });
    }

    public void showTimerListEmptyNotify() {
        tvTimerListEmptyNotify.setVisibility(View.VISIBLE);
    }

    public void hideTimerListEmptyNotify() {
        tvTimerListEmptyNotify.setVisibility(View.GONE);
    }


    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.e(TAG, "onStop: Đã dừng lại");
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (!OpenCVLoader.initDebug()) {
            OpenCVLoader.initAsync(OpenCVLoader.OPENCV_VERSION, this, mLoaderCallback);
        } else {
            mLoaderCallback.onManagerConnected(LoaderCallbackInterface.SUCCESS);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (mOpenCvCameraView != null) {
            mOpenCvCameraView.disableView();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mOpenCvCameraView != null)
            mOpenCvCameraView.disableView();
        Log.e(TAG, "onDestroy: Đã kết thúc.");
    }

    @Override
    public void onCameraViewStarted(int width, int height) {
        mRgba = new Mat(height, width, CvType.CV_8UC4);
        mRgbaF = new Mat(height, width, CvType.CV_8UC4);
        mRgbaT = new Mat(width, width, CvType.CV_8UC4);
    }

    @Override
    public void onCameraViewStopped() {
        mRgba.release();
    }

    @Override
    public Mat onCameraFrame(CameraBridgeViewBase.CvCameraViewFrame inputFrame) {
        try {
            mRgba = inputFrame.rgba();
            final Bitmap bmp = Bitmap.createBitmap(mRgba.cols(), mRgba.rows(), Bitmap.Config.ARGB_8888);
            Utils.matToBitmap(mRgba, bmp);
            if (ivImageCaptured.getVisibility() == View.VISIBLE) {
                // Nếu đang hiển thị khung nhìn cho 1 cam duy nhất
                ivImageCaptured.post(new Runnable() {
                    @Override
                    public void run() {
                        ivImageCaptured.setImageBitmap(bmp);
                    }
                });
            } else if (ivImageCapturedS1.getVisibility() == View.VISIBLE) {
                // Nếu có cho phép hiển thị cho cam 1 trong 4 cam
                ivImageCapturedS1.post(new Runnable() {
                    @Override
                    public void run() {
                        ivImageCapturedS1.setImageBitmap(bmp);
                    }
                });
            }
            VideoTransfer.send(bmp);
        } catch (Exception e) {
            //e.printStackTrace();
        }
        return mRgba; // This function must return
    }


}

File: app/src/main/java/vn/vistark/dknncy/ui/dashboard_screen/DashboardScreenActivity.kt
package vn.vistark.dknncy.ui.dashboard_screen

import android.annotation.SuppressLint
import android.app.ProgressDialog
import android.graphics.Bitmap
import android.os.Bundle
import android.util.Log
import android.view.SurfaceView
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import cn.pedant.SweetAlert.SweetAlertDialog
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import es.dmoral.toasty.Toasty
import kotlinx.android.synthetic.main.activity_dashboard_screen.*
import org.opencv.android.*
import org.opencv.android.CameraBridgeViewBase.CvCameraViewFrame
import org.opencv.android.CameraBridgeViewBase.CvCameraViewListener2
import org.opencv.core.CvType
import org.opencv.core.Mat
import vn.vistark.dknncy.R
import vn.vistark.dknncy.data.arduino_community.ArduinoInit
import vn.vistark.dknncy.data.firebase.FirebaseConfig
import vn.vistark.dknncy.data.modal.Auto
import vn.vistark.dknncy.data.modal.CurrentDetail
import vn.vistark.dknncy.data.modal.DefaultTimerItem
import vn.vistark.dknncy.data.modal.TimerItem
import vn.vistark.dknncy.ui.dashboard_screen.timer_dialog.TimerDialog
import vn.vistark.dknncy.ui.dashboard_screen.timer_list.TimerListAdapter
import vn.vistark.dknncy.utils.CommonUtils
import vn.vistark.dknncy.utils.NetworkUtils
import vn.vistark.dknncy.utils.ScreenUtils
import vn.vistark.dknncy.utils.TimeUtils
import java.text.SimpleDateFormat
import java.util.*
import kotlin.collections.ArrayList

class DashboardScreenActivity : AppCompatActivity(), CvCameraViewListener2 {
    private val timerItems = ArrayList<TimerItem>()
    lateinit var adapter: TimerListAdapter
    private lateinit var noInternetDialog: ProgressDialog

    private var establishStartAt: Long = 0

    // These variables are used (at the moment) to fix camera orientation from 270degree to 0degree
    var mRgba: Mat? = null

    //    var mRgbaF: Mat? = null
//    var mRgbaT: Mat? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_dashboard_screen)
        ScreenUtils.hideTitleBarAndTransparentStatusBar(this)

        hideTimerListEmptyNotify()

        noInternetDialog = CommonUtils.showNoInternetDialog(this)
        establishTime()
        runningCurrentTime()

        initEvents()
        initAutoSync()

        initRecyclerView()

        initCamera()

        ArduinoInit.asyncCommunity(this)

        checking()
    }

    // Hiển thị hộp thoại xác nhận thoát hay không
    fun powerOffConfirm() {
        val exitConfirm = SweetAlertDialog(this, SweetAlertDialog.WARNING_TYPE)
        exitConfirm.titleText = "ĐÓNG ỨNG DỤNG?"
        exitConfirm.confirmText = "Đóng"
        exitConfirm.cancelText = "Hủy"
        // Bắt sự kiện khi người dùng nhấn đồng ý đóng
        exitConfirm.setConfirmClickListener { finish() }
        // Bắt sự kiện khi người dùng hủy đóng
        exitConfirm.setCancelClickListener { sweetAlertDialog -> sweetAlertDialog.dismiss() }
        //
        try {
            exitConfirm.show()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun changeListByStateOfAutoPower(state: Boolean) {
        FirebaseConfig.autoRef.child(Auto.powerKey).setValue(state)
        if (state) {
            DefaultTimerItem.timerPortA.isPower = state
            DefaultTimerItem.timerPortB.isPower = state
            DefaultTimerItem.timerPortC.isPower = state
            DefaultTimerItem.timerPortD.isPower = state
            FirebaseConfig.updateData(DefaultTimerItem.timerPortA)
            FirebaseConfig.updateData(DefaultTimerItem.timerPortB)
            FirebaseConfig.updateData(DefaultTimerItem.timerPortC)
            FirebaseConfig.updateData(DefaultTimerItem.timerPortD)
        }

    }

    private fun initAutoSync() {
        Auto().apply {
            sync(FirebaseConfig.myRef, this@DashboardScreenActivity)
            powerSync(this@DashboardScreenActivity)
        }
    }

    @SuppressLint("SimpleDateFormat")
    private fun runningCurrentTime() {
        val simpleDateFormat = SimpleDateFormat("HH:mm:ss")
        Timer().schedule(object : TimerTask() {
            override fun run() {
                val date = Date(System.currentTimeMillis())
                updateCurrentTimeView(simpleDateFormat.format(date))
            }
        }, 1000, 1000)
    }

    // Cập nhật và hiển thị giờ đã khởi chạy
    private fun establishTime() {
        establishStartAt = System.currentTimeMillis()
        Timer().schedule(object : TimerTask() {
            override fun run() {
                val establishedTime = TimeUtils.millisToCounter(System.currentTimeMillis() - establishStartAt)
                updateEstablishTimeView(establishedTime)
                runOnUiThread {
                    if (NetworkUtils.isNetworkConnected(this@DashboardScreenActivity)) {
                        if (noInternetDialog.isShowing) {
                            noInternetDialog.dismiss()
//                            if (VideoTransfer.init()) {
//                                //
//                            } else {
//                                Toasty.error(this@DashboardScreenActivity, "Kết nối đến máy chủ hình ảnh thất bại, hãy khởi động lại", Toasty.LENGTH_SHORT, false).show()
//                            }
                        }
                    } else {
                        if (noInternetDialog.isShowing) {
                            noInternetDialog.show()
                        }
                    }
                }
            }
        }, 1000, 1000)
    }

    private val mLoaderCallback: BaseLoaderCallback = object : BaseLoaderCallback(this) {
        override fun onManagerConnected(status: Int) {
            if (status == LoaderCallbackInterface.SUCCESS) {
                (mOpenCvCameraView as CameraBridgeViewBase).enableView()
                showMultiCamera()
            } else {
                super.onManagerConnected(status)
            }
        }
    }

    private fun initCamera() {
        (mOpenCvCameraView as CameraBridgeViewBase).visibility = SurfaceView.VISIBLE
        (mOpenCvCameraView as CameraBridgeViewBase).setCvCameraViewListener(this)
    }

    private fun initRecyclerView() {
        rvTimerList.setHasFixedSize(true)
        rvTimerList.layoutManager = LinearLayoutManager(this)

        adapter = TimerListAdapter(timerItems)

        rvTimerList.adapter = adapter

        // lắng nghe danh sách hẹn giờ
        FirebaseConfig.timerRef.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                // This method is called once with the initial value and again
                // whenever data at this location is updated.
                DefaultTimerItem.check(dataSnapshot)
                for (ds in dataSnapshot.children) {
                    val ti = TimerItem().parse(ds) ?: return
                    updateTimerItem(ti)
                }
            }

            override fun onCancelled(error: DatabaseError) {
                // Failed to read value
                Log.w(TAG, "Lỗi khi đọc dữ liệu.", error.toException())
            }
        })

        // Xóa phần tử
        adapter.onLongClick = { ti ->
            if (ti.isDefault()) {
                Toasty.warning(this, "Không thể xóa hẹn giờ mặc định", Toasty.LENGTH_LONG, false).show()
            } else {
                // Nếu không phải là timer mặc định
                val sweetAlertDialog = SweetAlertDialog(this, SweetAlertDialog.WARNING_TYPE)
                sweetAlertDialog.setTitleText("XÁC NHẬN XÓA HẸN GIỜ?").show()
                sweetAlertDialog.confirmText = "Xóa"
                sweetAlertDialog.cancelText = "Đóng"
                sweetAlertDialog.setConfirmClickListener { s ->
                    FirebaseConfig.timerRef.child(ti.id.toString() + "").removeValue()
                    timerItems.remove(ti)
                    adapter.notifyDataSetChanged()
                    if (timerItems.indexOf(ti) < 0) {
                        Toasty.success(this, "Đã xóa").show()
                    }
                    s.dismiss()
                }
                sweetAlertDialog.setCancelClickListener { s -> s.dismiss() }
                sweetAlertDialog.show()
            }
        }

        // Sửa phần tử
        adapter.onClick = { ti ->
            TimerDialog.showEdit(this, ti)
        }
    }

    private fun updateTimerItem(timerItem: TimerItem) {
        val index = timerItems.indexOfFirst { x -> x.id == timerItem.id }
        if (index >= 0) {
            if (timerItems[index].isDiff(timerItem))
                timerItems[index].update(timerItem)
        } else {
            timerItems.add(timerItem)
        }
        adapter.notifyDataSetChanged()
    }


    fun showMultiCamera() {
        lnMultiCameraButton.visibility = View.GONE
        ivImageCaptured.visibility = View.GONE
        lnMultiCamLayout.visibility = View.VISIBLE
    }

    fun chooseShowCam1() {
        lnMultiCameraButton.visibility = View.VISIBLE
        ivImageCaptured.visibility = View.VISIBLE
        lnMultiCamLayout.visibility = View.GONE
    }

    fun initEvents() {
        // Bắt sự kiện khi nhấn vào nút đóng
        ivAppPowerOffButton.setOnClickListener { ivAppPowerOffButton.post { powerOffConfirm() } }

        // Bắt sự kiện khi nhấn vào nút thêm mới hẹn giờ
        btnAddNewTimer.setOnClickListener { TimerDialog.showAdd(this@DashboardScreenActivity) }

        // Bắt sự kiện khi nhấn vào nút nhiều camera
        lnMultiCameraButton.setOnClickListener { showMultiCamera() }

        // Sự kiện khi nhấn vào 1 khung hình của cam bất kỳ trong chế độ xem nhiều camera
        // Nhấn màn hình 1
        ivImageCapturedS1.setOnClickListener { chooseShowCam1() }
        // Nhấn màn hình 2,3,4
        val ivs = arrayOf(ivImageCapturedS2, ivImageCapturedS3, ivImageCapturedS4)
        for (v in ivs) {
            v.setOnClickListener { v.post { Toasty.info(this@DashboardScreenActivity, "Không nhận được tín hiệu từ camera này!", Toasty.LENGTH_SHORT, false).show() } }
        }

        // Sự kiện cho nút nguồn auto
        scTimerAutoPower.setOnCheckedChangeListener { compoundButton, b -> changeListByStateOfAutoPower(b) }
    }


    override fun onBackPressed() {
        powerOffConfirm()
    }

    fun updateAutoHumidity(start: String, end: String) {
        tvAutoHumidity.post { tvAutoHumidity.text = "$start - $end%" }
    }

    fun updateAutoTemp(start: String, end: String) {
        tvAutoTemp.post { tvAutoTemp.text = "$start - $end" }
    }

    val stateAutoPower: Boolean
        get() = scTimerAutoPower.isChecked

    fun updateTimerAutoPower(state: Boolean?) {
        scTimerAutoPower.post { scTimerAutoPower.isChecked = state ?: false }
    }

    fun updateHumidityView(humidity: String) {
        tvHumidity.post { tvHumidity.text = "$humidity%" }
    }

    fun updateTemperature(temperature: String) {
        tvTemperature.post { tvTemperature.text = "$temperature°C" }
    }

    fun updateCurrentTimeView(currentTime: String?) {
        tvCurrentTime.post { tvCurrentTime.text = currentTime }
    }

    fun updateEstablishTimeView(establedTime: String?) {
        tvEstablishTime.post { tvEstablishTime.text = establedTime }
    }

    fun showTimerListEmptyNotify() {
        tvTimerListEmptyNotify.visibility = View.VISIBLE
    }

    fun hideTimerListEmptyNotify() {
        tvTimerListEmptyNotify.visibility = View.GONE
    }

    override fun onResume() {
        super.onResume()
        if (!OpenCVLoader.initDebug()) {
            OpenCVLoader.initAsync(OpenCVLoader.OPENCV_VERSION, this, mLoaderCallback)
        } else {
            mLoaderCallback.onManagerConnected(LoaderCallbackInterface.SUCCESS)
        }
    }

    override fun onPause() {
        super.onPause()
        (mOpenCvCameraView as CameraBridgeViewBase).disableView()
    }

    override fun onDestroy() {
        super.onDestroy()
        (mOpenCvCameraView as CameraBridgeViewBase).disableView()
        Log.e(TAG, "onDestroy: Đã kết thúc.")
    }

    override fun onCameraViewStarted(width: Int, height: Int) {
        mRgba = Mat(height, width, CvType.CV_8UC4)
//        mRgbaF = Mat(height, width, CvType.CV_8UC4)
//        mRgbaT = Mat(width, width, CvType.CV_8UC4)
    }

    override fun onCameraViewStopped() {
        mRgba?.release()
    }

    override fun onCameraFrame(inputFrame: CvCameraViewFrame): Mat? {
        try {
            mRgba = inputFrame.rgba()
            val bmp = Bitmap.createBitmap(mRgba!!.cols(), mRgba!!.rows(), Bitmap.Config.ARGB_8888)
            println("Hiển thị BITMAP size ${mRgba?.size()?.width}")
            Utils.matToBitmap(mRgba, bmp)
            if (ivImageCaptured.visibility == View.VISIBLE) {
                println("Hiển thị 1 cam")
                // Nếu đang hiển thị khung nhìn cho 1 cam duy nhất
                ivImageCaptured.post { ivImageCaptured.setImageBitmap(bmp) }
            } else if (ivImageCapturedS1.visibility == View.VISIBLE) {
                println("Hiển thị nhiều cam")
                // Nếu có cho phép hiển thị cho cam 1 trong 4 cam
                ivImageCapturedS1.post { ivImageCapturedS1.setImageBitmap(bmp) }
            }
//            VideoTransfer.send(bmp)
        } catch (e: Exception) {
            e.printStackTrace()
            //e.printStackTrace();
        }
        return mRgba // This function must return
    }

    companion object {
        private val TAG = DashboardScreenActivity::class.java.simpleName
    }

    private fun checking() {
        Timer().schedule(object : TimerTask() {
            override fun run() {
                checkStateUpdateAndSendData()
            }
        }, 1000, 1000)
    }

    private fun checkStateUpdateAndSendData() {
        // Lặp trong danh sách các cổng
        for (i in timerItems.indices) {
            val previous = timerItems[i]
            if (scTimerAutoPower.isChecked) {
                // Nếu chế độ tự động được bật, những timer thủ công sẽ được bỏ qua
                if (!timerItems[i].isDefault()) continue

                // Nếu không phải là những timer thủ công, tiến hành kiếm tra điều kiện và bật/tắt
                val currHumidity = CurrentDetail.humidity.toIntOrNull() ?: return
                val currTemperature = CurrentDetail.temperature.toIntOrNull() ?: return

                // Máy bơm
                if (timerItems[i].isPump())
                    timerItems[i].isState =
                            (currHumidity < Auto.humidityStartValue ||
                                    currTemperature > Auto.tempEndValue) &&
                                    timerItems[i].isPower

                // Đèn
                else if (timerItems[i].isLight())
                    timerItems[i].isState =
                            (currTemperature < Auto.tempStartValue ||
                                    currHumidity > Auto.humidityEndValue) &&
                                    timerItems[i].isPower

                // Âm thanh
                else
                    timerItems[i].isState = timerItems[i].isPower
            } else {
                // Nếu chế độ tự động tắt, tiến hành tắt các timer tự động
                if (timerItems[i].isDefault()) {
                    println("${timerItems[i].id} là mặc định")
                    timerItems[i].isState = false
                    timerItems[i].isPower = false
                } else {
                    // Sau đó kiểm tra khung giờ bình thường
                    timerItems[i].isState =
                            TimeUtils.isInTimer(timerItems[i].start, timerItems[i].end) &&
                                    timerItems[i].isPower
                }
            }

            ArduinoInit.sendCommand(timerItems[i].port, timerItems[i].isState)
            FirebaseConfig.updateData(timerItems[i]) // -> Tự động sẽ được cập nhật
            if (previous.isDiff(timerItems[i])) {
                Log.d(TAG, "checkStateUpdateAndSendData: Đã cập nhập")
            }
            // Tiến hành gửi thông tin xuống arduino
        }
    }

}